File: node/src/sensor/vibration.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a vibration sensor.
 *
 * @internal
 */
export enum VibrationProperty {
  /** Whether vibration is detected. */
  Detected = 'detected',
}

/**
 * Property values of a vibration sensor.
 *
 * @internal
 */
export interface VibrationSensorProperties {
  [VibrationProperty.Detected]: boolean;
}

/** Read-only proxy interface for a vibration sensor. */
export interface VibrationSensorLike extends SensorLike {
  readonly type: SensorType.Vibration;
  readonly onPropertyChanged: Observable<PropertyChangeOf<VibrationSensorProperties>>;

  getValue(property: VibrationProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Vibration sensor. */
export class VibrationSensor<TStorage extends object = Record<string, any>> extends Sensor<VibrationSensorProperties, TStorage> {
  readonly type = SensorType.Vibration;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Vibration Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [VibrationProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report vibration detection state.
   *
   * @param value - True when vibration is currently detected.
   *
   * @example
   * ```ts
   * vibration.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [VibrationProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link VibrationSensor}. */
export const vibrationMeta = defineSensor({
  type: SensorType.Vibration,
  category: SensorCategory.Sensor,
  assignmentKey: 'vibration',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [VibrationProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: VibrationProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [VibrationProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: VibrationProperty.Detected,
    commandProperty: VibrationProperty.Detected,
    deviceClass: 'vibration',
  },
});

File: python/camera_ui_sdk/sensor/vibration.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class VibrationProperty(StrEnum):
    """Property names of a vibration sensor."""

    Detected = "detected"
    """Whether vibration is detected."""


class VibrationSensorProperties(TypedDict):
    """Property values of a vibration sensor."""

    detected: bool


class VibrationPropertyChangeData(TypedDict):
    """Property change payload emitted on VibrationSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a VibrationProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class VibrationSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a vibration sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Vibration

    @property
    def onPropertyChanged(self) -> Observable[VibrationPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[VibrationProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class VibrationSensor(Sensor[VibrationSensorProperties, TStorage, str], Generic[TStorage]):
    """Vibration sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Vibration Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({VibrationProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Vibration

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report vibration detection state.

        Args:
            value: True when vibration is currently detected.

        Example:
            ```python
            vibration.setDetected(True)
            ```
        """
        self._write_state({VibrationProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
